### PR TITLE
Support gradient accumulation using Horovod's `backward_passes_per_step`

### PR DESCRIPTION
**Pull request by krshrimali**
## What does this PR do?

This PR attempts to fix https://github.com/PyTorchLightning/pytorch-lightning/issues/11732.

Uses `backward_passes_per_step` kwarg in Horovod's `DistributedOptimizer` for the purpose as mentioned in the issue.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Added
 
+- Enable gradient accumulation using Horovod's `backward_passes_per_step`
```
Comment (rohitgr7) [inline CHANGELOG.md:12]:
```suggestion
- Enable gradient accumulation using Horovod's `backward_passes_per_step` ([#11911](https://github.com/PyTorchLightning/pytorch-lightning/pull/11911))
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -9,6 +9,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
...
+
+
 - Add new `DETAIL` log level to provide useful logs for improving monitoring and debugging of batch jobs
```
Comment (rohitgr7) [inline CHANGELOG.md:15]:
```suggestion
- Add new `DETAIL` log level to provide useful logs for improving monitoring and debugging of batch jobs ([#11008](https://github.com/PyTorchLightning/pytorch-lightning/pull/11008))
```
looks like we missed this... adding it here.

File: pytorch_lightning/strategies/horovod.py [hunk 3]
```diff
@@ -184,8 +196,13 @@ def post_backward(self, closure_loss: torch.Tensor) -> None:
 
     def _wrap_optimizers(self, optimizers: List[Optimizer]) -> List["hvd.DistributedOptimizer"]:
         """Wraps optimizers to perform gradient aggregation via allreduce."""
+        accumulate_grad_batches = self.trainer.accumulate_grad_batches
```
Comment (rohitgr7) [inline pytorch_lightning/strategies/horovod.py:199]: can you add a test?

File: CHANGELOG.md (hunk 1 above)
Comment (krshrimali) [reply CHANGELOG.md:12]: Thanks, @rohitgr7 - sorry, I missed this.

File: pytorch_lightning/strategies/horovod.py (hunk 3 above)
Comment (krshrimali) [reply pytorch_lightning/strategies/horovod.py:199]: Sure, makes sense. Getting a weird error with `horovod.torch` not available from the helper function `_module_available` in utils. Working on it. :)

Comment (krshrimali) [reply pytorch_lightning/strategies/horovod.py]: Tried adding a test for both CPU and GPU devices, with `accumulate_grad_batches` as an argument in the constructor to trainer. Also minor fix, as it should be `self.lightning_module.trainer` instead of `self.trainer`. I had to explicitly set `_HOROVOD_AVAILABLE` as `True` to run tests locally, for some reason `_module_available` returns `False` for `horovod.torch`.

File: tests/models/test_horovod.py [hunk 4]
```diff
@@ -82,6 +82,23 @@ def test_horovod_cpu(tmpdir):
...
+        weights_save_path=str(tmpdir),
+        gradient_clip_val=1.0,
+        enable_progress_bar=False,
+        max_epochs=1,
+        limit_train_batches=0.4,
+        limit_val_batches=0.2,
+        accumulate_grad_batches=4,
+        strategy="horovod",
+    )
+    _run_horovod(trainer_options)
```
Comment (krshrimali) [inline tests/models/test_horovod.py:99]: Should these (and the other added test in this PR) be separate tests? Or can we just add `accumulate_grad_batches=4` to existing tests? I thought a separate test is better as it will help debug failures related to this change only, keeping other tests untouched.

File: pytorch_lightning/strategies/horovod.py [hunk 5]
```diff
@@ -112,6 +118,12 @@ def _unpack_lightning_optimizer(opt):
         for optimizer in optimizers:
             hvd.broadcast_optimizer_state(optimizer, root_rank=0)
 
+        accumulation_scheduler = self.lightning_module.trainer.accumulation_scheduler
```
Comment (ananthsub) [inline pytorch_lightning/strategies/horovod.py:121]: trainer is an argument to this function already
```suggestion
        accumulation_scheduler = trainer.accumulation_scheduler
```

File: pytorch_lightning/strategies/horovod.py [hunk 6]
```diff
@@ -184,8 +196,13 @@ def post_backward(self, closure_loss: torch.Tensor) -> None:
 
     def _wrap_optimizers(self, optimizers: List[Optimizer]) -> List["hvd.DistributedOptimizer"]:
         """Wraps optimizers to perform gradient aggregation via allreduce."""
+        accumulate_grad_batches = self.lightning_module.trainer.accumulate_grad_batches
```
Comment (ananthsub) [inline pytorch_lightning/strategies/horovod.py:199]: pass accumulate_grad_batches as an argument to this function from the `setup` function above. it's easier to test this function in isolation then

Comment (krshrimali) [reply pytorch_lightning/strategies/horovod.py]: Thanks, @ananthsub for taking a look. I have made the changes, but curious to know what you feel about the tests added in this PR?

File: pytorch_lightning/strategies/horovod.py [hunk 7]
```diff
@@ -112,7 +118,13 @@ def _unpack_lightning_optimizer(opt):
         for optimizer in optimizers:
             hvd.broadcast_optimizer_state(optimizer, root_rank=0)
 
-        self.optimizers = self._wrap_optimizers(optimizers)
+        accumulation_scheduler = trainer.accumulation_scheduler
+        if accumulation_scheduler.epochs != [0]:
+            raise MisconfigurationException(
```
Comment (ananthsub) [inline pytorch_lightning/strategies/horovod.py:123]: Please add a test for this

File: tests/models/test_horovod.py [hunk 8]
```diff
@@ -131,6 +148,24 @@ def test_horovod_multi_gpu(tmpdir):
...
+    trainer_options = dict(
+        default_root_dir=str(tmpdir),
+        weights_save_path=str(tmpdir),
+        gradient_clip_val=1.0,
+        enable_progress_bar=False,
+        max_epochs=1,
+        limit_train_batches=0.4,
+        limit_val_batches=0.2,
+        accumulate_grad_batches=4,
+        gpus=2,
```
Comment (ananthsub) [inline tests/models/test_horovod.py:163]:
```suggestion
        accelerator="gpu",
        devices=2,
```

File: pytorch_lightning/strategies/horovod.py (hunk 7 above)
Comment (krshrimali) [reply pytorch_lightning/strategies/horovod.py:123]: Thanks, @ananthsub for the suggestion. I've added the tests, but instead of using `_run_horovod`, I went for `trainer.fit(model)` since `_run_horovod` dumps the `dict` into a string object (and does not return the keys as integers, so the test will fail for this reason). I hope the newly added tests look good now.

File: tests/models/test_horovod.py (hunk 8 above)
Comment (krshrimali) [reply tests/models/test_horovod.py]: Done! I also took liberty to change other instances in this file, I hope that's okay?

File: tests/models/test_horovod.py [hunk 9]
```diff
@@ -125,12 +164,54 @@ def test_horovod_multi_gpu(tmpdir):
...
+        enable_progress_bar=False,
+        max_epochs=1,
+        limit_train_batches=0.4,
+        limit_val_batches=0.2,
+        accumulate_grad_batches={0: 4, 2: 2},
+        accelerator="gpu",
+        devices=2,
+        strategy="horovod",
+    )
+    with pytest.raises(MisconfigurationException):
```
Comment (carmocca) [inline tests/models/test_horovod.py:211]: Always match an informative piece of the expected exception or warning


```suggestion
    with pytest.raises(MisconfigurationException, match="Horovod.*does not support.*accumulate_grad_batches"):
```

File: tests/models/test_horovod.py [hunk 10]
```diff
@@ -125,12 +164,54 @@ def test_horovod_multi_gpu(tmpdir):
...
+        default_root_dir=str(tmpdir),
+        weights_save_path=str(tmpdir),
+        gradient_clip_val=1.0,
+        enable_progress_bar=False,
+        max_epochs=1,
+        limit_train_batches=0.4,
+        limit_val_batches=0.2,
+        accumulate_grad_batches={0: 4, 2: 2},
+        accelerator="gpu",
+        devices=2,
```
Comment (carmocca) [inline tests/models/test_horovod.py:208]: This test does not need to run on 2 devices.

```suggestion
        devices=1,
```

File: tests/models/test_horovod.py [hunk 11]
```diff
@@ -125,12 +164,54 @@ def test_horovod_multi_gpu(tmpdir):
...
+    Strategy on multi-gpus."""
+    model = ClassificationModel()
+    trainer = Trainer(
+        default_root_dir=str(tmpdir),
+        weights_save_path=str(tmpdir),
+        gradient_clip_val=1.0,
+        enable_progress_bar=False,
+        max_epochs=1,
+        limit_train_batches=0.4,
+        limit_val_batches=0.2,
```
Comment (carmocca) [inline tests/models/test_horovod.py:205]: Since we expect to fail on `setup`, we don't need this

```suggestion
```

File: tests/models/test_horovod.py [hunk 12]
```diff
@@ -125,12 +164,54 @@ def test_horovod_multi_gpu(tmpdir):
...
+
+@RunIf(min_gpus=2, skip_windows=True, horovod_nccl=True)
+def test_horovod_multi_gpu_accumulate_grad_batches_different(tmpdir):
+    """Ensure MisConfigurationException for different `accumulate_grad_batches` at different epochs for Horovod
+    Strategy on multi-gpus."""
+    model = ClassificationModel()
+    trainer = Trainer(
+        default_root_dir=str(tmpdir),
+        weights_save_path=str(tmpdir),
+        gradient_clip_val=1.0,
```
Comment (carmocca) [inline tests/models/test_horovod.py:201]: This should be irrelevant

```suggestion
        default_root_dir=tmpdir,
```

File: tests/models/test_horovod.py [hunk 13]
```diff
@@ -125,12 +164,54 @@ def test_horovod_multi_gpu(tmpdir):
...
+        enable_progress_bar=False,
+        max_epochs=1,
+        limit_train_batches=0.4,
+        limit_val_batches=0.2,
+        accumulate_grad_batches={0: 4, 2: 2},
```
Comment (carmocca) [inline tests/models/test_horovod.py:206]: A simpler value would also raise, no?

```suggestion
        accumulate_grad_batches=2,
```

File: tests/models/test_horovod.py [hunk 14]
```diff
@@ -125,12 +164,54 @@ def test_horovod_multi_gpu(tmpdir):
...
+        accelerator="gpu",
```
Comment (carmocca) [inline tests/models/test_horovod.py:207]: Instead of adding two tests, one for GPU and one for CPU, you could do

```suggestion
        accelerator="auto",
```

so it is chosen based on the available environment

**Review by carmocca — COMMENTED** [review]
Always try to keep tests as minimal as possible

File: tests/models/test_horovod.py (hunk 13 above)
Comment (krshrimali) [reply tests/models/test_horovod.py:206]: Thanks, @carmocca for the question - I think this test needs different batches at different epochs for it to fail, hence the dictionary here. Does this answer your question?

**Comment by krshrimali** [comment]
Thanks, @carmocca for the review and suggestions. Just to let you know, I'll make the changes to merge the cpu and gpu tests into one as you suggested, and will update this PR again. I've committed through your suggestions for some of your comments, and others will come along with my next commit. Thank you for the suggestions again! :))

**Review by ananthsub — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
LGTM!

As a side note, we should open an issue about rethinking Horovod testing.
